server/app/services/apikey/ApiKeyService.java

```diff
@@ -0,0 +1,291 @@
+
+    if (subnetString.isBlank()) {
+      return form.withError(FORM_FIELD_NAME_SUBNET, "Subnet cannot be blank.");
+    }
+
+    if (subnetString.endsWith("0") && banGlobalSubnet) {
+      return form.withError(FORM_FIELD_NAME_SUBNET, "Subnet cannot allow all IP addresses.");
+    }
+
+    try {
+      new SubnetUtils(subnetString);
+    } catch (IllegalArgumentException e) {
+      return form.withError(FORM_FIELD_NAME_SUBNET, "Subnet must be in CIDR notation.");
+    }
+
+    apiKey.setSubnet(subnetString);
+
+    return form;
+  }
+
+  // Pattern for matching and extracting form field names that specify
+  // granting read permission for a program.
+  // These field names have the format "grant-program-read[program-slug]"
+  // Where "program-slug" is the sluggified name of the program the key should
+  // be granted read access for e.g. "grant-program-read[utility-discount-program]".
+  private static final Pattern GRANT_PROGRAM_READ_PATTERN =
+      Pattern.compile("^grant-program-read\\[([\\w\\-]+)\\]$");
+
+  private DynamicForm resolveGrants(DynamicForm form, ApiKey apiKey)
+      throws ProgramNotFoundException {
+    ImmutableSet<String> programSlugs = programService.getAllProgramSlugs();
+
+    for (String formDataKey : form.rawData().keySet()) {
+      Matcher matcher = GRANT_PROGRAM_READ_PATTERN.matcher(formDataKey);
+
+      if (!matcher.find()) {
+        continue;
+      }
+
+      matcher.matches();
+
+      String programSlug = matcher.group(1);
+
+      if (!programSlugs.contains(programSlug)) {
+        throw new ProgramNotFoundException(programSlug);
+      }
+
+      apiKey.getGrants().grantProgramPermission(programSlug, Permission.READ);
+    }
+
+    return form;
+  }
+
+  private String generateSecret(int length) {
+    KeyGenerator keyGen;
+
+    try {
+      keyGen = KeyGenerator.getInstance("AES");
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+
+    keyGen.init(length);
+
+    byte[] secret = keyGen.generateKey().getEncoded();
+    return Base64.getEncoder().encodeToString(secret);
+  }
+
+  private String salt(String message) {
+    byte[] rawMessage = Base64.getDecoder().decode(message);
+    byte[] rawKey = Base64.getDecoder().decode(secretSalt);
+
+    HashFunction hashFunction = Hashing.hmacSha256(rawKey);
+    HashCode saltedMessage = hashFunction.hashBytes(rawMessage);
+
+    return Base64.getEncoder().encodeToString(saltedMessage.asBytes());
+  }
+
+  private String getAuthorityId(CiviFormProfile profile) {
+    String authorityId = profile.getAuthorityId().join();
+
+    if (authorityId != null) {
+      return authorityId;
+    }
+
+    if (environment.isDev()) {
+      return "dev-admin";
+    }
+
+    throw new RuntimeException("ApiKey creator must have authority_id.");
+  }
+
+  /** Holds state relevant to the result of attempting to create an {@link ApiKey}. */
+  public static class ApiKeyCreationResult {
+    private final Optional<ApiKey> apiKey;
```

Comment:
[nit] it's nice to put these in the same order as the constructor params

Reply:
done